kani-compiler/src/codegen_cprover_gotoc/codegen/foreign_function.rs

```diff
@@ -0,0 +1,155 @@
+    fn codegen_ffi_shim(&mut self, shim_name: InternedString, instance: Instance<'tcx>) -> Stmt {
+        debug!(?shim_name, ?instance, sym=?self.symbol_table.lookup(shim_name), "generate_foreign_shim");
+
+        let loc = self.codegen_span(&self.tcx.def_span(instance.def_id()));
+        let unsupported_check = self.codegen_ffi_unsupported(instance, loc);
+        Stmt::block(vec![unsupported_check], loc)
+    }
+
+    /// Generate type for the given foreign instance.
+    fn codegen_ffi_type(&mut self, instance: Instance<'tcx>) -> Type {
+        let fn_name = self.symbol_name(instance);
+        let fn_abi = kani_middle::fn_abi(self.tcx, instance);
+        let loc = self.codegen_span(&self.tcx.def_span(instance.def_id()));
+        let params = fn_abi
+            .args
+            .iter()
+            .enumerate()
+            .filter_map(|(idx, arg)| {
+                (!arg.is_ignore()).then(|| {
+                    let arg_name = format!("{fn_name}::param_{idx}");
+                    let base_name = format!("param_{idx}");
+                    let arg_type = self.codegen_ty(arg.layout.ty);
+                    let sym = Symbol::variable(&arg_name, &base_name, arg_type.clone(), loc)
+                        .with_is_parameter(true);
+                    self.symbol_table.insert(sym);
+                    arg_type.as_parameter(Some(arg_name.into()), Some(base_name.into()))
+                })
+            })
+            .collect();
+        let ret_type = self.codegen_ty(fn_abi.ret.layout.ty);
+
+        if fn_abi.c_variadic {
+            Type::variadic_code(params, ret_type)
+        } else {
+            Type::code(params, ret_type)
+        }
+    }
+
+    /// Kani does not currently support FFI functions except for built-in CBMC functions.
+    ///
+    /// This will behave like `codegen_unimplemented_stmt` but print a message that includes
+    /// the name of the function not supported and the calling convention.
+    pub fn codegen_ffi_unsupported(&mut self, instance: Instance<'tcx>, loc: Location) -> Stmt {
+        let fn_name = &self.symbol_name(instance);
+        debug!(?fn_name, ?loc, "codegen_ffi_unsupported");
+
+        // Save this occurrence so we can emit a warning in the compilation report.
+        let entry = self.unsupported_constructs.entry("foreign function".into()).or_default();
+        entry.push(loc);
+
+        let call_conv = kani_middle::fn_abi(self.tcx, instance).conv;
+        let msg = format!("call to foreign \"{call_conv:?}\" function `{fn_name}`");
+        let url = if call_conv == Conv::C {
+            "https://github.com/model-checking/kani/issues/2423"
+        } else {
+            "https://github.com/model-checking/kani/issues/new/choose"
+        };
+        self.codegen_assert_assume(
+            Expr::bool_false(),
+            PropertyClass::UnsupportedConstruct,
+            &GotocCtx::unsupported_msg(&msg, Some(url)),
+            loc,
+        )
```

Comment:
I assume this will be reported as an assertion failure with the same source location of the unsupported function, correct?